add new feature encodeRadioDetails and optimize sprintfPrettyDetails

dstroy0 opened the pull request:
encodeRadioDetails will encode an array of 10 uint32_t  (40 bytes) with the exact same information that is found in printPrettyDetails.

decodeRadioDetails works in the same manner as sprintfPrettyDetails, but you pass it the encoded array and it formats the output buffer for you

This is useful for anyone needing status information with no output and stringent memory use requirements.

encodeRadioDetails is meant to be tailored to your use case, you can output as much or as little as you want by adding new members or by commenting existing members out

Decode manually or with decodeRadioDetails.

Documentation needs updating to reflect behavior.

2bndy5 commented:
@dstroy0 Please enable all workflows on your fork. Typically this can be done by going to https://github.com/dstroy0/RF24/actions, but I can't confirm this because I don't have admin privileges for your fork (nor do I want admin rights for your fork). Please also check the repo settings for your fork at https://github.com/dstroy0/RF24/settings/actions I just want to be sure that the changes you just submitted after opening the PR fixed the PicoSDK and Linux builds.

dstroy0 commented:
> @dstroy0 Please enable all workflows on your fork. Typically this can be done by going to https://github.com/dstroy0/RF24/actions, but I can't confirm this because I don't have admin privileges for your fork (and I do not want admin rights for your fork). Please also check the repo settings for your fork at https://github.com/dstroy0/RF24/settings/actions I just want to be sure that the changes you just submitted after opening the PR fixed the PicoSDK and Linux builds.

I enabled all of the workflows.  I can see them but there's no run workflow button.

2bndy5 commented:
You can't trigger them manually because they're not setup for that (a rather advanced use case).

You should be able to trigger the Linux & PicoSDK workflows by making a commit that edits the lib source code somehow; I usually just add a bogus comment or new line somewhere to trigger them. 

The Arduino CI and PlatformIO CI workflows are triggered by any commit involving the examples folder (not examples_linux or examples_pico folders). Again, I usually just add a bogus comment or new line somewhere to trigger them.

2bndy5 commented:
Judging from the latest runs on your fork, the Linux builds and Pico SDK builds look ok again ✅

2bndy5 commented:
Fun fact: I setup the docs CI to upload built docs (folder of HTML files) as [workflow artifacts](https://github.com/dstroy0/RF24/actions/runs/1736922223#artifacts) which you can download and open on your local machine for inspection.

dstroy0 commented:
> Judging from the latest runs on your fork, the Linux builds and Pico SDK builds look ok again ✅

Excellent, forgetting to remove that debug output before issuing the pr was a big oof.

dstroy0 commented:
could maybe add the functions to KEYWORD 2 so the arduino ide highlights them in orange

2bndy5 commented:
> could maybe add the functions to KEYWORD 2 so the arduino ide highlights them in orange

No objections there. In fact, I didn't check if the same was applied to `sprintfPrettyDetails()`.

2bndy5 commented on examples/encodeRadioDetails/encodeRadioDetails.ino:
```diff
@@ -0,0 +1,225 @@
+*/
+
+/**
+   A simple example of sending data from 1 nRF24L01 transceiver to another.
```
```suggestion
   A simple example of sending debug info from 1 nRF24L01 transceiver to another.
```

2bndy5 commented:
I have a long-term goal of making a executable hardware test that checks all publicly exposed functions, and I think this new example would be better suited to that task. Generally, I like to port Arduino examples to Linux and PicoSDK when applicable...

2bndy5 commented on examples/encodeRadioDetails/encodeRadioDetails.ino:
```diff
@@ -0,0 +1,225 @@
+      Serial.print(F(" bytes on pipe "));
+      Serial.print(pipe);                     // print the pipe number
+      Serial.println(F(": "));
+      for (int i = 0; i < 10; i++) {
```
Would it be more transparent to use `bytes` var instead of hard coding this loop to `10` iterations?

dstroy0 replied:
I'm not sure what you mean.

2bndy5 replied:
```cpp
/*
  For this example, we'll be using a payload containing
  radio details encoded with RF24::encodeRadioDetails()
  it is sent by the sender and decoded by the receiver
  using decodeRadioDetails(), you can manually decode
  your encoded details by printing out the payload value
  in binary by or blinking a led for each bit in the
  payload
*/
uint32_t payload[10] = {0};
```
Looks like all payloads in this example are 10 bytes long.

```cpp
  // save on transmission time by setting the radio to only transmit the
  // number of bytes we need to transmit
  radio.setPayloadSize(sizeof(payload)); // float datatype occupies 4 bytes
```
The comment here should now describe the payload as a byte array (&not a float value).

```cpp
    uint8_t pipe;
    if (radio.available(&pipe)) {             // is there a payload? get the pipe number that recieved it
      uint8_t bytes = radio.getPayloadSize(); // get the size of the payload
```
Will this `bytes` var always be equal to 10? Seems like it should.

If so, `bytes` should probably be used to iterate over the payload when printing out the binary representation. By saying "transparent", I mean that users will be able to read/understand that the entire payload is iterated (& not just the first arbitrary 10 bytes).

dstroy0 replied:
I see, you are correct.  I think I should also change the serial output to reflect that it's a "binary representation of the payload" or something.

2bndy5 replied:
yeah, don't want to let them think they chose the wrong baudrate for `Serial.begin()`

2bndy5 replied:
I just realized that each element in the payload is a `uint32_t`... 10 uint32_t is not 10 bytes. I think it comes out to be 40 actual bytes which is greater than the max payload size?! Am I calculating that wrong?

dstroy0 replied:
it is 40 bytes, we can make it less than 40 by getting rid of a couple of members or by splitting the transmit in the loop, I forgot we aren't using network!

2bndy5 replied:
The radio's Mac payload length is 32. This is a hardware limit not a lib limit.

2bndy5 replied:
If using the networking layers, then the max payload length is 24 without message fragmentation (a feature specific to RF24Network). This is due to needed header info. Fragmentation is disabled for ATTiny devices by default (enabled for all others).

dstroy0 replied:
I split the payload in the example, each half is 24 bytes

2bndy5 commented:
A quick glance at changes in RF24.cpp make me think there's room for improvement 
- like only using 1 bit for a bool instead of a whole byte
- Most of the radio's registers don't actually use bit 7 for anything
- etc, etc.

Simply put I need more time to review this. I need to return to fixing the CMake solution for other RF24 libs for now.

dstroy0 commented:
> A quick glance at changes in RF24.cpp make me think there's room for improvement
> 
>     * like only using 1 bit for a bool instead of a whole byte
> 
>     * Most of the radio's registers don't actually use bit 7 for anything
> 
>     * etc, etc.
> 
> 
> Simply put I need more time to review this. I need to return to fixing the CMake solution for other RF24 libs for now.

I understand.  Boolean are represented by 1 bit inside of the bit array, let me know if I can help with anything in the meantime.

Off the top of my head, csn_pin and ce_pin don't need to be uint16_t unless you have a use case where the pin number is greater than 255, one of the uint16_t that gets transmitted can be a uint8_t,, 3 byte savings.

In real use, I know the csn_pin, and ce_pin so I don't send those.

dstroy0 commented:
> @TMRh20 Was there a specific reason (like a particular chip) to switch to uint16_t for the pin numbers? I think ManiacBug originally used uint8_t for this.
> 
> EDIT: disregard my question... Found [the commit that changed the pin numbers to uint16_t](https://github.com/nRF24/RF24/commit/eb056706e9fba23fe8f2c97c8b46727f15ef8d8c). Apparently, it was to accommodate a device that used an Intel Atom CPU called the [Minnowboard](https://www.mouser.com/ProductDetail/Silicom/MBT-4220?qs=sg81DGxMojVwoTpGiNGBpQ%3D%3D).
> 
> @dstroy0 So, for full compatibility with all supported devices, I'd say the pin numbers should be uint16_t when encoded into the byte array.

Ok, the other candidate is ARD, we can just send the uint8_t register value  in encodeRadioDetails and expand it in decodeRadioDetails.  Do you have any ideas to make fragmenting and reassembly in the example clearer?

TMRh20 commented:
I remember somebody requested uint16_t for PIN numbers a long while back. Can’t remember what boards though

> On Jan 24, 2022, at 7:51 AM, dstroy0 ***@***.***> wrote:
> 
> ﻿
> Another idea I had was to switch the encoded_details bit array from uint32_t to uint8_t, what do you think of that? Then we can really allocate only the minimum necessary.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

2bndy5 commented:
> Do you have any ideas to make fragmenting and reassembly in the example clearer?

Yep. My first reaction to that approach was to send a byte that indicates which half of the payload is sent (similar to how RF24Network does it).

> candidate is ARD, we can just send the uint8_t register

Agreed. The OBSERVE_TX register is the same situation.

I also like your idea about using a uint8_t buffer.

Additionally
- CONFIG, RF_SETUP, & FEATURE registers need only a half byte each.
- EN_AA & EN_RXADDR registers only needs 6 bits each
- RX_PW_P0 register (synonymously used for statically sized payloads on all pipes in this lib) only uses 6 bits also
- SETUP_AW register only needs 2 bits

---
I've just about finished up with fixing the CMake problems we found. Think I'll get some sleep and dive into this PR later.

2bndy5 commented:
oh yeah, the dynamic payloads feature could be represented with 1 bit since this lib doesn't support configuring the DYNPD register for each pipe. Although, that register should always correspond with bit 2 in the FEATURE register.

dstroy0 commented:
I will make the new put methods for those bit sizes later today.

dstroy0 commented:
I’m making templates for the bit manipulation methods, I think having so many methods is more confusing than having a template would be.

dstroy0 commented:
need some sort of guard when decoding and referencing the enums in case there's a problem with the bit order it will print incorrect information instead of causing panic

dstroy0 commented:
> need some sort of guard when decoding and referencing the enums in case there's a problem with the bit order it will print incorrect information instead of causing panic

if the message is encoded incorrectly or decoded incorrectly it's possible that the enum indexes can be greater than the number of enum members

2bndy5 commented:
Please excuse my arrogance here, but that problem seems specific to customizing what info is included in the  buffer.

Enums aren't actually arrays, rather (I think) they're just typical variables. So by saying "enum index", I assume you mean enumerator's value. None of our enum values need more than 2 bits (if that helps) - `RF24_PA_ERROR` is a relic that isn't used anywhere.

dstroy0 commented:
Not a problem, I’m not communicating this correctly.  It’s the same type of fault as accessing nonexistent array members.  What is happening is when the string is getting formatted in sprintf_P it gets something other than expected as an argument.  Instead of pointing to one of the string literals it points to something else outside their memory and breaks sprintf.  I think anyways.

> On Jan 24, 2022, at 5:20 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> Please excuse my arrogance here, but that problem seems specific to customizing what info is included in the buffer.
> 
> Enums aren't actually arrays, rather (I think) they're just typical variables. So by saying "enum index", I assume you mean enumerator's value. None of our enum values need more than 2 bits (if that helps) - RF24_PA_ERROR is a relic that isn't used anywhere.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

2bndy5 commented:
What I still don't understand is how an incorrect value could be introduced into the buffer. From what I know about sprintf(), there is no way to guard against out-of-boundary problems, so I'm trying to focus on making sure the correct data is where it should be in the buffer. And, I don't see an obvious point of error in the encoding/decoding functions.

dstroy0 commented:
There aren’t any, they could be introduced by users trying to customize the function.  All that I need to do is constrain those sprintf arguments that access those string literals so they can’t go out of bounds, inside of encodeRadioDetails right before sprintf_P.  Then it will print incorrect info after the error instead of breaking.

2bndy5 commented:
I started looking at these changes, and did some chores while I was inspecting. None of the changes should actually affect functionality.
- changed the class `bit_manipulation_methods` into a namespace called `BitManip`. This way you don't need to instantiate an object just to access encapsulated methods. Note this class could've been a struct since all methods are public, but I think using a namespace is more memory friendly.
- renamed the `BitManip` methods to typical getter/setter names. The args are well named enough to describe where data is going to/from.
- removed useless forward declaration in RF24.h
- removed a bunch of redundant casting and scoping specifiers
- made use of the defined macro `_BV` in some cases. `_BV` comes from Arduino land and is defined as `1 << (x)`.

If you look at the Linux CI logs (specifically [in the "build lib" step](https://github.com/dstroy0/RF24/runs/4929062702?check_suite_focus=true#step:9:6)), you were getting some pedantic warnings being raised:
```
[ 16%] Building CXX object CMakeFiles/rf24.dir/RF24.cpp.o
/home/runner/work/RF24/RF24/RF24.cpp: In member function ‘void RF24::encodeRadioDetails(uint8_t*)’:
/home/runner/work/RF24/RF24/RF24.cpp:1133:78: warning: ‘*’ in boolean context, suggest ‘&&’ instead [-Wint-in-bool-context]
 1133 |     temp_bool = (static_cast<bool>(read_register(FEATURE) & _BV(EN_DYN_ACK)) * 2);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/home/runner/work/RF24/RF24/RF24.cpp:1136:78: warning: ‘*’ in boolean context, suggest ‘&&’ instead [-Wint-in-bool-context]
 1136 |     temp_bool = (static_cast<bool>(read_register(FEATURE) & _BV(EN_ACK_PAY)) * 1);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```
I removed the useless arithmetic referenced in these warnings.

dstroy0 commented:
I'm working on removing the bloat comments and I added a primer into encodeRadioDetails for people who want to change things.  I think that since the max spi speed supported by the radio is 10MHz we can use 4 or 5 bits to represent it after it is casted to a uint8_t, I still haven't went through and changed any of the bit sizes, I wanted to get the comments making sense first before I started poking it again.

2bndy5 commented:
The new methods could use some docs as well, but that usually gets added last after the API is finalized. Having those comments up-to-date will help. Hint: Doxygen will add comments found within the definition of a member to it's corresponding docs.
```cpp
/**
 * @brief a brief description
 *
 * A longer description. The second `*` at the
 * beginning of this comment block tells doxygen
 * to pay attention to this comment.
 * 
 * @tparam T A description for the for template parameter
 * @param paramName paramName description.
 * @returns No return descriptions are needed for
 * functions that return void.
 */
template<typename T>
bool func(T paramName)
{
    /** 
     * This gets added to the long description
     * because the comment starts with 2 `*`
     */
}
```

2bndy5 commented:
@dstroy0 You can probably see now why I wanted those CI workflows getting triggered. 😉

dstroy0 commented:
> @dstroy0 You can probably see now why I wanted those CI workflows getting triggered. 😉

Yes it's very handy.

2bndy5 commented on RF24.cpp:
```diff
@@ -985,6 +1016,608 @@ void RF24::sprintfPrettyDetails(char *debugging_information)
+    // 34 forty_bit_register_array[4]
+    BitManip::packValue(forty_bit_register_array[4], 8, encoded_details, &encoded_details_index, &bit_index);
+
+    // pipe 1 40 bit address register (static_cast<uint8_t>(RX_ADDR_P0 + 1))
+    temp_8_bit = (static_cast<uint8_t>(RX_ADDR_P0 + 1));
+    arrayify_address_register(forty_bit_register_array, temp_8_bit);
+    // 35 forty_bit_register_array[0]
+    BitManip::packValue(forty_bit_register_array[0], 8, encoded_details, &encoded_details_index, &bit_index);
+    // 36 forty_bit_register_array[1]
+    BitManip::packValue(forty_bit_register_array[1], 8, encoded_details, &encoded_details_index, &bit_index);
+    // 37 forty_bit_register_array[2]
+    BitManip::packValue(forty_bit_register_array[2], 8, encoded_details, &encoded_details_index, &bit_index);
+    // 38 forty_bit_register_array[3]
+    BitManip::packValue(forty_bit_register_array[3], 8, encoded_details, &encoded_details_index, &bit_index);
+    // 39 forty_bit_register_array[4]
+    BitManip::packValue(forty_bit_register_array[4], 8, encoded_details, &encoded_details_index, &bit_index);
+
+    // pipe 2 8 bit address register (static_cast<uint8_t>(RX_ADDR_P0 + 2)
+    temp_8_bit = (static_cast<uint8_t>(RX_ADDR_P0 + 2));
+    arrayify_byte_register(eight_bit_register_array, temp_8_bit);
+    // 40 eight_bit_register_array[0]
+    BitManip::packValue(eight_bit_register_array[0], 8, encoded_details, &encoded_details_index, &bit_index);
+
+    // pipe 3 8 bit address register (static_cast<uint8_t>(RX_ADDR_P0 + 3)
+    temp_8_bit = (static_cast<uint8_t>(RX_ADDR_P0 + 3));
+    arrayify_byte_register(eight_bit_register_array, temp_8_bit);
+    // 41 eight_bit_register_array[0]
+    BitManip::packValue(eight_bit_register_array[0], 8, encoded_details, &encoded_details_index, &bit_index);
+
+    // pipe 4 8 bit address register (static_cast<uint8_t>(RX_ADDR_P0 + 4)
+    temp_8_bit = (static_cast<uint8_t>(RX_ADDR_P0 + 4));
+    arrayify_byte_register(eight_bit_register_array, temp_8_bit);
+    // 42 eight_bit_register_array[0]
+    BitManip::packValue(eight_bit_register_array[0], 8, encoded_details, &encoded_details_index, &bit_index);
+
+    // pipe 4 8 bit address register (static_cast<uint8_t>(RX_ADDR_P0 + 5)
+    temp_8_bit = (static_cast<uint8_t>(RX_ADDR_P0 + 5));
+    arrayify_byte_register(eight_bit_register_array, temp_8_bit);
+    // 43 eight_bit_register_array[0]
+    BitManip::packValue(eight_bit_register_array[0], 8, encoded_details, &encoded_details_index, &bit_index);
```
Anything against using a for loops?

dstroy0 replied:
No. It will be nice to look at, easier to read.  I'll push that out now.

2bndy5 commented on RF24.h:
```diff
@@ -1972,6 +2056,152 @@ class RF24 {
+     * @param bit_index The bitwise index of the bit array `encoded_details`.
+     */
+    template <typename T>
+    static void getBits(T *out, int number_of_bits, uint8_t *encoded_details, uint8_t *encoded_details_index, uint16_t *bit_index)
```
Would it be easier to make this return a value of type `T` instead of passing by reference? Most of the calls to this function allocate a new byte to pass `out` by reference, then pass `out`'s value to `sprintf()`. It could be called directly from within a `sprintf()` call... The only times that a temp byte needs to be allocated is when you perform arithmetic on the channel value (`temp_byte + 2400`) or use the value as an index to an array of translating strings.

2bndy5 commented on RF24.h:
```diff
@@ -1972,6 +2056,152 @@ class RF24 {
+                uint8_t index = (*bit_index) % 8;
+                SetBit(&encoded_details[*encoded_details_index], index);
+            }
+            (*bit_index)++;
+            if (*bit_index % 8 == 0)
+            {
+                (*encoded_details_index)++;
+            }
```
Repetitious code should live in a reusable function. Preferably `inline` in this case.

2bndy5 commented on examples/encodeRadioDetails/encodeRadioDetails.ino:
```diff
@@ -0,0 +1,265 @@
+      split_payload_2d_array[0][i] = payload[i]; // first 20 elements
+      split_payload_2d_array[1][i] = payload[i]; // second 20 elements
+    }
+    for (int i = 0; i < 2; i++) // iterate twice to transmit, because we split our payload into two pieces
```
Its more performant to use `++i` in for loops because of the difference in how pre/post `++` operators are overloaded.

dstroy0 replied:
> @dstroy0 put on the brakes!
> 
> The more I think about it, I'm having trouble getting behind these additions.
> 
>     1. Transmitting radio details to another nRF24 would require that the configuration of both radios should match. So, it doesn't make any practical sense to transmit debug info from 1 device to another.
>        
>        * I know you were thinking in terms of networking, but the configuration for network nodes is even more restricted to a certain criteria.
> 
>     2. These changes require a lot more code with very little in the way of optimization.
>        
>        * needs another buffer allocated to store debug info
>        * packing the debug data into another buffer then unpacking it into a char buffer seems redundant
>        * the CPU cycles spent on doing bit manipulation may actually outweigh the memory benefit of manipulating bits
>        * the added namespace of functions is likely increasing the compile size much more than the amount of memory being spared
> 
>     3. I think `sprintf()` could be used better because it returns the number bytes written to the specified buffer. This returned data could be used for subsequent calls to `sprintf()` because the number of bytes written does not include the terminating null byte that is automatically appended.
>        ```c++
>        uint16_t bytes_used = 0;
>        char* buf = new char[870];
>        bytes_used += sprintf(buf, "some %s data", "string");
>        sprintf(buf + bytes_used, "some number %d", 42);
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        
>        * I think the original `printPrettyDetails()` could be made to reuse `sprintfPrettyDetails()`, but this idea might not be any more efficient than using `printf()` directly.
> 
> 
> I mentioned this in your previous PR, but I'll say it again because we're still discussing the concepts of these changes. Generally, we discuss new additions/changes in a new [issue](https://github.com/nRF24/RF24/issues) before opening a PR and spending time revising code. I feel like I'm aimlessly reviewing these changes because I'm not sure what this PR is designed to satisfy.
> 
>     * sending me a personal email is not a good alternative because I'm 1 person in a community working to develop/maintain these libraries. Directing the conversation to only me excludes the rest of the people with vested interest in this library.

The output method encodeRadioDetails() uses much less ram and flash than other output methods available in the library when used on its own, that alone makes it worth pursuing.  I also don't understand the point of having two sprintf output methods either, it's there to make debugging encodeRadioDetails() easier and that's it right now, it's pointless.

2bndy5 reviewed (changes requested):
### 1<sup>st</sup> review
I mostly focused on conceptual stuff here. For the next review, I intend to focus on better code reuse since... well
![image](https://user-images.githubusercontent.com/14963867/151109180-ac38ecd4-e306-4dda-ace7-807299128cdd.png)
We need to work on this.

Also, I thought the original idea for these changes was to compliment `sprintfPrettyDetails()` (& friends), but I don't see that code taking advantage of these changes. Is my timing a bit premature?

dstroy0 commented on RF24.cpp:
```diff
@@ -985,6 +1016,608 @@ void RF24::sprintfPrettyDetails(char *debugging_information)
+    temp_8_bit = ((read_register(DYNPD) && (read_register(FEATURE) & _BV(EN_DPL))) * 1);
+    BitManip::packValue(temp_8_bit, 8, encoded_details, &encoded_details_index, &bit_index);
+    // 18 (static_cast<bool>(temp_8_bit & _BV(ENAA_P5)) + 48)
+    temp_8_bit = read_register(EN_AA);
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P5)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 19 (static_cast<bool>(temp_8_bit & _BV(ENAA_P4)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P4)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 20 (static_cast<bool>(temp_8_bit & _BV(ENAA_P3)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P3)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 21 (static_cast<bool>(temp_8_bit & _BV(ENAA_P2)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P2)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 22 (static_cast<bool>(temp_8_bit & _BV(ENAA_P1)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P1)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 23 (static_cast<bool>(temp_8_bit & _BV(ENAA_P0)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P0)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
```
Yes, I will work on that later those were for me.  It will.

> ### 1st review
> 
> I mostly focused on conceptual stuff here. For the next review, I intend to focus on better code reuse since... well ![image](https://user-images.githubusercontent.com/14963867/151109180-ac38ecd4-e306-4dda-ace7-807299128cdd.png) We need to work on this.
> 
> Also, I thought the original idea for these changes was to compliment `sprintfPrettyDetails()` (& friends), but I don't see that code taking advantage of these changes. Is my timing a bit premature?

Yes, I want to get rid of the sprintf in encodedRadioDetails, I think we should change the function declaration to have a default argument sprintfPrettyDetails(debugging_information, encoded_array=0).  That would get rid of quite a few lines, and increase maintainability because there aren't two format_string doing similar things, it wouldn't break anyone's existing project who decided to use the new feature.  I think that's a good approach, what do you think? There are a lot of other things I need to take care of that you pointed out so once I'm through with that I think we should be in a good place to change the output behavior once it's verified correct.

2bndy5 replied:
I think changing `decodeDetails()` into an overload of `sprintfPrettyDetails()` that takes 2 args instead of 1 would be a better approach.

I imagine it like so
```cpp
void sprintfPrettyDetails(char* debug_info) {
    // Create buffer using encodeDetails()
    uint8_t packedDets[38];
    encodeDetails(&packedDets);

    // Pass the buffer to the overloaded function
    sprintfPrettyDetails(debug_info, &packedDets);
}
```

dstroy0 replied:
> I think changing `decodeDetails()` into an overload of `sprintfPrettyDetails()` that takes 2 args instead of 1 would be a better approach.
> 
> I imagine it like so
> 
> ```c++
> void sprintfPrettyDetails(char* debug_info) {
>     // Create buffer using encodeDetails()
>     uint8_t packedDets[38];
>     encodeDetails(&packedDets);
> 
>     // Pass the buffer to the overloaded function
>     sprintfPrettyDetails(debug_info, &packedDets);
> }
> ```

`
sprintfPrettyDetails(char *debugging_information, uint8_t *encoded_details=0)
{
 if(encoded_details != 0)
  {
   //use encoded details for output
  }
 else
 {
  //read registers as normal
 }
 //output
}
decodeRadioDetails(char *debugging_information, uint8_t *encoded_details)
{
 sprintfPrettyDetails(debugging_information, encoded_details);
}
`

is what I was thinking of

dstroy0 replied:
I'm sorry for the bad code comment lol, I'm learning this as fast as I can.

dstroy0 replied:
so really, it ends up just being a wrapper for sprintfPrettyDetails.

2bndy5 replied:
Github uses [markdown syntax](https://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax). Use ["fenced" code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) to preserve whitespace.

2bndy5 replied:
> `if(encoded_details != 0)`

This is a classic NULL vs 0 problem. I was trying to avoid this by using function overloads.

dstroy0 commented:
OT: I got the clang format extension working in windows vscode, LLVM has the clang-format binary on their website

dstroy0 commented on RF24.cpp:
```diff
@@ -985,6 +1016,608 @@ void RF24::sprintfPrettyDetails(char *debugging_information)
+     */
+    uint16_t bit_index = 0;                         // bitwise index of encoded_details
+    uint8_t encoded_details_index = 0;              // encoded_details array index
+    bool temp_bool = false;                         // arithmetic performed on this before passing as argument to bit manipulation method
+    uint8_t temp_8_bit = 0;                         // arithmetic performed on this before passing as argument to bit manipulation method    
```
done

2bndy5 commented:
@dstroy0 put on the brakes!

The more I think about it, I'm having trouble getting behind these additions.
1. Transmitting radio details to another nRF24 would require that the configuration of both radios should match. So, it doesn't make any practical sense to transmit debug info from 1 device to another.
   - I know you were thinking in terms of networking, but the configuration for network nodes is even more restricted to a certain criteria.
2. These changes require a lot more code with very little in the way of optimization.
   - needs another buffer allocated to store debug info
   - packing the debug data into another buffer then unpacking it into a char buffer seems redundant
   - the CPU cycles spent on doing bit manipulation may actually outweigh the memory benefit of manipulating bits
3. I think `sprintf()` could be used better because it returns the number bytes written to the specified buffer. This returned data could be used for subsequent calls to `sprintf()` because the number of bytes written does not include the terminating null byte that is automatically appended.
   ```cpp
   uint8_t bytes_used = 0;
   char* buf = new char[870];
   bytes_used = sprintf(buf, "some %s data", "string");
   sprintf(buf + bytes_used, "some number %d", 42);
   ```

I mentioned this in your previous PR, but I'll say it again because we're still discussing the concepts of these changes. Generally, we discuss new additions/changes in a new [issue](https://github.com/nRF24/RF24/issues) before opening a PR and spending time revising code. I feel like I'm aimlessly reviewing these changes because I'm not sure what this PR is designed to satisfy.
   - sending me a personal email is not a good alternative because I'm 1 person in a community working to develop/maintain these libraries. Directing the conversation to only me excludes the rest of the people with vested interest in this library.

2bndy5 commented:
> when used on its own

This is the part I'm confused about. What scenario would benefit from using `encodeRadioDetails()` on its own. What's preventing this method from being private/protected?

If customizing the function would pretty much mean re-writing it, then make it protected. That way people could customize it via inheritance.

I have no problem replacing the first implementation of `sprintfPrettyDetails()`, but I'm not sure we need to expose the packed buffer that this new implementation would be using.

dstroy0 commented:
I tried compiling as is, this is what I got for avr boards:

examples/GettingStarted
GettingStarted.ino
no output
7278 flash
251 ram
printDetails()
10888 flash
267 ram
printPrettyDetails()
11602 flash
267 ram
sprintfPrettyDetails()
12266 flash
261 ram
encodeRadioDetails()
9410 flash
261 ram
decodeRadioDetails()
13128 flash
251 ram
encodeRadioDetails() + decodeRadioDetails()
15028 flash
261 ram

2bndy5 commented:
> sprintfPrettyDetails()
12266 flash
261 ram

38% flash on atmega328P (+4% more than printDets; +2% more than printPrettyDets)

vs 

> encodeRadioDetails() + decodeRadioDetails()
15028 flash
261 ram

46% flash on atmega328P

With 32KB of flash on atmega328P, this would take almost half of that (& no reduction of ram). I don't think people should be using sprintfDets on devices that can use printf, but we have had requests in the past to use printDets on a non-primary Serial bus (for which sprintfDets could satisfy). I expect people using SAMD-based processors (which includes nRF5xxxx) need to use sprintfDets because printf is no longer a part of the Arduino Unified API (established in 2016).

Using flash might be faster execution than reading from ram, but there's the balance between execution and allocation that should be considered. A library shouldn't hog up so much of either if the users application is going to have to compromise when debugging development.

dstroy0 commented:
> > sprintfPrettyDetails()
> > 12266 flash
> > 261 ram
> 
> 38% flash on atmega328P (+4% more than printDets; +2% more than printPrettyDets)
> 
> vs
> 
> > encodeRadioDetails() + decodeRadioDetails()
> > 15028 flash
> > 261 ram
> 
> 46% flash on atmega328P
> 
> With 32KB of flash on atmega328P, this would take almost half of that (& no reduction of ram). I don't think people should be using sprintfDets on devices that can use printf, but we have had requests in the past to use printDets on a non-primary Serial bus (for which sprintfDets could satisfy). I expect people using SAMD-based processors (which includes nRF5xxxx) need to use sprintfDets because printf is no longer a part of the Arduino Unified API (established in 2016).
> 
> Using flash might be faster execution than reading from ram, but there's the balance between execution and allocation that should be considered. A library shouldn't hog up so much of either if the user's application is going to have to compromise when debugging development.

Right, so if I were on that I platform wouldn't use either, I would use encodeRadioDetails and send that information to be decoded by a more robust platform or I would decode it manually OR I could make an app to decode the output.  That was the idea anyway, to get as much of the information out in the most information dense package possible.

2bndy5 commented:
>I would use encodeRadioDetails and send that information to be decoded by a more robust platform

This requires a working configuration such that a wireless connection is established. Most of time when people use debugging information, they're trying to figure out why their app doesn't establish a wireless connection.

>OR I could make an app to decode the output.

Speaking in terms of a wide-spread ADD occurring due to ubiquitous tech (namely the millennial generation), nobody is going to want to write a decoding algorithm just to fit debugging functionality into their app. If you meant having a dedicated example that does the decoding, then that might be more viable. However, I still suspect people would copy-paste such example code into their problematic app (at which point flash mem would be saturated again).

dstroy0 commented:
In those cases, printing out the binary value of encoded_details and manually decoding it still puts out more information and uses less flash and ram than the other output methods.  I also have a problem with it using sprintf, decodeRadioDetails feels like an afterthought because it was, I’m not attached to it at all.

> On Jan 26, 2022, at 7:03 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> I would use encodeRadioDetails and send that information to be decoded by a more robust platform
> 
> This requires a working configuration such that a wireless connection is established. Most of time when people use debugging information, they're trying to figure out why their app doesn't establish a wireless connection.
> 
> OR I could make an app to decode the output.
> 
> Speaking in terms of a wide-spread ADD occurring due to ubiquitous tech (namely the millennial generation), nobody is going to want to write a decoding algorithm just to fit debugging functionality into their app. If you meant having a dedicated example that does the decoding, then that might be more viable. However, I still suspect people would copy-paste such example code into their problematic app (at which point flash mem would be saturated again).
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

2bndy5 commented:
> printing out the binary value of encoded_details and manually decoding it still puts out more information and uses less flash and ram than the other output methods

ok, I'm sold on exposing encodeDets(). Personally, I'd probably decode it by copy-paste the hex output into a python script running in a separate terminal.

dstroy0 commented:
> > printing out the binary value of encoded_details and manually decoding it still puts out more information and uses less flash and ram than the other output methods
> 
> ok, I'm sold on exposing encodeDets(). Personally, I'd probably decode it by copy-paste the hex output into a python script running in a separate terminal.

yes, exactly!

2bndy5 commented:
Ironically, the python script I'd be using would be modified from the CirPy lib's `RF24.print_details()` (which outputs everything + kitchen sink) 😆

If this is the intention of encodeDets(), then we can probably do away with bit manipulation and preserve the registers' values. That should slim down the compile size even more.

As for decodeDets vs sprintfDets, I'm also ok with ditching decodeDets and focusing on optimizing sprintfDets (if possible).

dstroy0 commented:
> Ironically, the python script I'd be using would be modified from the CirPy lib's `RF24.print_details()` (which outputs everything + kitchen sink) 😆
> 
> If this is the intention of encodeDets(), then we can probably do away with bit manipulation and preserve the registers' values. That should slim down the compile size even more.
> 
> As for decodeDets vs sprintfDets, I'm also ok with ditching decodeDets and focusing on optimizing sprintfDets (if possible).

Yes, I have ideas for sprintfPrettyDetails to reduce compile size, the fastest way would be to make duplicate PSTR const char*

2bndy5 commented:
BTW you don't have to quote my entire post when replying. It tends to do a number on my mouse's scroll wheel (and I just got a new one) as the tread gets really long.

dstroy0 commented:
Ok cool!

dstroy0 commented:
Am I reading that function right, 0-23,28,29? and 10,11,16 are the 40 bit registers?

dstroy0 commented:
we can use arrayify_address_register to grab the 40 bit registers I think that should stay and I will cut the other functions and draft a c++ version of that python function

2bndy5 commented:
> 10,11,16 are the 40 bit registers?

yes.

Inclusively: 0-9, 12-15, 17-23, 28-29 are single byte registers

2bndy5 commented:
It would help for readability if you used the macros from [nRF24L01.h](https://github.com/nRF24/RF24/blob/master/nRF24L01.h), but I would increment in loops when possible.

dstroy0 commented:
It should be easy to document what registers are being fetched in the comments

dstroy0 commented:
for (uint8_t i = 0; i < 0x1E; ++i) {
    if (i <= 9) {
      // single byte
    }
    // 10 40 bit
    // 11 40 bit
    if (i >= 12 && i <= 15) {
      // single byte
    }
    // 16 40 bit
    if (i >= 17 && i <= 23) {
      // single byte
    }
    if (i == 28) {
      // single byte
    }
    if (i == 29) {
      // single byte
    }
  }

dstroy0 commented:
How do I use code tags properly?

2bndy5 commented:
https://github.com/nRF24/RF24/pull/825#discussion_r793061422

2bndy5 commented:
### here ya go!
```cpp
void dump_registers(uint8_t end = FEATURE + 1) {
    for (uint8_t i = NRF_CONFIG; i < end; ++i) {
        if (i == RX_ADDR_P0 || i == RX_ADDR_P1 || i == TX_ADDR)
            // get 40-bit registers
        else if (i != 0x18 && i != 0x19 && i != 0x1a && i != 0x1b) // skip undocumented registers
            // get single byte registers
    }
}
```

2bndy5 commented:
I don't think you'll need `arrayifyAddrReg()`. You can simply
```cpp
uint8_t out[38];
uint8_t addrBuf[5];
reg_read(i, &addrBuf);
memcpy(&out + i, &addrBuf, 5);
```
Or simply pass the output buffer + index offset to `reg_read()`
```
uint8_t out[38];
// `i` won't always be equal to `index_offset` since we're skipping some undocumented registers
reg_read(i, &out + index_offest);
```

2bndy5 commented:
ok I went ahead and pushed my implementation of encodeDets without use of BitManip and arrayify*() (which consequently have been removed along with decodeDets).

I tested my modifications using the new example (stripped away everything from `loop()` as well) and verified that the newly added print_details.py script works as expected (located in the new example's folder). However, I tested the example on my ItsyBitsy RP2040 because that's what I have wired up right now, and it doesn't support printPrettyDets (nor the older printDets).

dstroy0 commented:
I've just tested avr, adafruit M4, teensy 4.1, and espressif esp32 and esp8266, all working.  Excellent!

dstroy0 commented:
with the radio object instantiated and using sprintfPrettyDetails global memory use is down to 232 bytes on avr

2bndy5 commented:
I'm still brainstorming here, but I think there's a way to pass the debug_info buffer directly to sprintf_reg_addr() instead of allocating a new char buffer and inserting that into the debug_info buffer.

If work it properly, the same could be done for the contextual format of the auto-ack info also.

dstroy0 commented:
I tried a few different methods to see if I could get rid of repetitive string literals in format_string being stored in flash but they all use more program space and slightly more memory.  Each method rebuilt format_string out of separate string literals.

2bndy5 commented:
What I'm thinking is several calls to sprintf() that make use of the returned value (number of bytes written to debug_info buffer).
1. sprintf all the debug_info up until auto-ack data
2. sprintf the auto-ack data
3. sprintf "Primary Mode" and just the TX addr label
4. sprintf_addr_reg TX addr
5. for loop through the pipe info
   1. sprintf pipe label (including "open"/closed" status)
   2. sprintf address data for pipes. This would call sprintf_addr_reg() for pipes 0 & 1. Then use sprintf for pipes 2-5.
      - the ulterior intention is to also get rid of sprintf_reg_byte()

dstroy0 commented:
> What I'm thinking is several calls to sprintf() that make use of the returned value (number of bytes written to debug_info buffer).
> 
>     1. sprintf all the debug_info up until auto-ack data
> 
>     2. sprintf the auto-ack data
> 
>     3. sprintf "Primary Mode" and just the TX addr label
> 
>     4. sprintf_addr_reg TX addr
> 
>     5. for loop through the pipe info
>        
>        1. sprintf pipe label (including "open"/closed" status)
>        2. sprintf address data for pipes. This would call sprintf_addr_reg() for pipes 0 & 1. Then use sprintf for pipes 2-5.
>           
>           * the ulterior intention is to also get rid of sprintf_reg_byte()
> 
> 
> In the end, I want to stop allocating additional buffers local to sprintfDets and just keep dumping data directly to the debug_info buffer via sprintf().

yes that was one way that I tried, it uses a few bytes of memory every call to sprintf_P

2bndy5 commented:
> yes that was one way that I tried, it uses a few bytes of memory every call to sprintf_P

were you saving the return number to variable or just ignoring it?

2bndy5 commented:
you can commit anything and git will allow us to "revert" an entire commit of changes

dstroy0 commented:
I used it to increment the buffer, it was set with every return.

> On Jan 27, 2022, at 5:35 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> yes that was one way that I tried, it uses a few bytes of memory every call to sprintf_P
> 
> were you saving the return number to variable or just ignoring it?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

2bndy5 commented:
hmm, what if we just do step 5?

dstroy0 commented:
I think step 5 might yield useful results, every buffer has that null term occupying space.

dstroy0 commented:
I tried strcat() puts() with separate string literals, this was the slowest and largest out of the methods I tried.

2bndy5 commented:
I added the CE & CSN pin numbers to the encodeDet() as well as SPI speed or'd with `isPlusVariant << 4`... Total mandatory bytes for the buffer is 43 now. I also updated the docs about encodeDets and converted the comment within the function's definition into a MD (MarkDown) table:
![image](https://user-images.githubusercontent.com/14963867/151489597-e38b43c5-826f-45f5-8062-82513ac6a1b3.png)

2bndy5 commented:
Ran a local test about my idea for sprintfDets:
### as is (without multiple calls to sprintf)
```
Sketch uses 8396 bytes (27%) of program storage space. Maximum is 30720 bytes.
Global variables use 275 bytes (13%) of dynamic memory, leaving 1773 bytes for local variables. 
```

### with multiple calls to sprintf (& no additionally allocated local buffers)
```
Sketch uses 7968 bytes (25%) of program storage space. Maximum is 30720 bytes.
Global variables use 279 bytes (13%) of dynamic memory, leaving 1769 bytes for local variables. Maximum is 2048 bytes.
```

Looks like flash goes down by 428 bytes, but ram goes up by 4 bytes. This is an acceptable balance for me. I'll start cleaning up the tested code and commit, so you can play around a bit further (which might be better than me trying to explain what I did).

sprintfDets and sprintf_reg_addr now return the number of bytes that were altered in the given buffer (just like sprintf does).

2bndy5 commented on RF24.h:
```diff
@@ -106,11 +106,11 @@ typedef enum {
     RF24_CRC_16
 } rf24_crclength_e;
 
+
```
```suggestion
```

2bndy5 commented on RF24.h:
```diff
@@ -1972,7 +1979,6 @@ class RF24 {
 
 };
 
-
```
```suggestion

```

2bndy5 commented on RF24.cpp:
```diff
@@ -820,170 +804,146 @@ void RF24::printPrettyDetails(void) {
+            read_register(i, encoded_details, 5);
+            encoded_details += 5;
+        }
+        else if (i != 0x18 && i != 0x19 && i != 0x1a && i != 0x1b){ // skip undocumented registers
```
```suggestion
       else if (i != 0x18 && i != 0x19 && i != 0x1a && i != 0x1b) { // skip undocumented registers
```

2bndy5 commented on RF24.cpp:
```diff
@@ -820,170 +804,146 @@ void RF24::printPrettyDetails(void) {
+        (char *)(pgm_read_ptr(&rf24_feature_e_str_P[
+            static_cast<bool>(read_register(FEATURE) & _BV(EN_ACK_PAY)) * 1])),
+        (char *)(pgm_read_ptr(&rf24_feature_e_str_P[
+            (read_register(DYNPD) && (read_register(FEATURE) & _BV(EN_DPL))) * 1]))
+    );
```
```suggestion
            (read_register(DYNPD) && (read_register(FEATURE) & _BV(EN_DPL))) * 1])));
```

dstroy0 commented:
Everything looks great! I wont have a chance to test anything until this evening.  I will figure out how to make the clang-format extension use that format specifier.

dstroy0 commented:
I got clang-format to use the .clang-format file

2bndy5 commented:
```
Sketch uses 7958 bytes (25%) of program storage space. Maximum is 30720 bytes.
Global variables use 275 bytes (13%) of dynamic memory, leaving 1773 bytes for local variables. Maximum is 2048 bytes.
```
YES! That's what I was going for: Reduced flash, and don't add to ram. @dstroy0 Thanks for fixing my mistakes.

2bndy5 commented:
I just had another itch to re-use the lib's private SPI buffers (`spi_rxbuff` and `spi_txbuff`) in `*print*_address_register()`. This idea would benefit all printDets methods...

2bndy5 commented:
Nope nevermind. I forgot that those private SPI buffers only exist on platforms that have more memory
```c
#if defined (RF24_LINUX) || defined (XMEGA_D3) || defined (RF24_RP2)
```

2bndy5 reviewed (approved):
I'm ready to merge this unless you can think of more optimizations

dstroy0 commented:
Everything looks great to me, I’ve been using encodeRadioDetails and sprintfRadioDetails as part of some of my projects startup status dump and they work as intended.  I’ll be sure to open an issue in the future before doing a PR.  This was a lot of fun.

> On Jan 29, 2022, at 8:07 PM, Brendan ***@***.***> wrote:
> 
> ﻿
> @2bndy5 approved this pull request.
> 
> I'm ready to merge this unless you can think of more optimizations
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

2bndy5 commented:
> I’ll be sure to open an issue in the future before doing a PR

Yes. Please do as that would provide a way to have separate discussions about ideas and code review
